### PR TITLE
[Review] Enable matrix::copyRows for row major input

### PR DESCRIPTION
**Pull request by tfeher**
Enable matrix::copyRows operation for row_major input. This will be used in SVM to enable row major input https://github.com/rapidsai/cuml/issues/2198

We use a helper function prim `cache:get_vecs()` from cuML. Because of this, the [cache_util.cuh](https://github.com/rapidsai/cuml/blob/branch-0.19/cpp/src_prims/cache/cache_util.cuh) file is copied to raft (only the namespace is updated). A separate PR shall update the the rest of the cache function and the associated tests and move it to RAFT.

### CODE REVIEW

**Comment by tfeher** [comment]
rerun tests

**Comment by cjnolet** [comment]
@tfeher, the Linkage gtest assertion failure should be fixed when #178 is merged. I'm still investigating timeouts in cuda 10.2. Don't have any leads on that yet, unfortunately.

**Review by cjnolet — APPROVED** [review]
LGTM. Mostly just small typos and I don't think those need to hold up the merge if they aren't fixed right away.

File: cpp/include/raft/cache/cache_util.cuh [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+
+/**
+ * @brief Assign cache location to a set of keys using LRU replacement policy.
+ *
+ * The keys and the corresponding cache_set arrays shall be sorted according
+ * to cache_set in ascending order. One block should be launched for every cache
+ * set.
+ *
+ * Each cache set is sorted according to time_stamp, and values from keys
+ * are filled in starting at the oldest time stamp. Enties that were accessed
```
[the remark this answers is not in the document]
Comment (tfeher) [reply cpp/include/raft/cache/cache_util.cuh:233]: Corrected.

File: cpp/include/raft/cache/cache_util.cuh [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+ * at the current time are not reassigned.
+ *
+ * @tparam nthreads number of threads per block
+ * @tparam assaciativity number of keys in a cache set
```
[the remark this answers is not in the document]
Comment (tfeher) [reply cpp/include/raft/cache/cache_util.cuh:237]: Fixed.

File: cpp/include/raft/cache/cache_util.cuh [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+  int block_offset = blockIdx.x * associativity;
+
+  const int items_per_thread = raft::ceildiv(associativity, nthreads);
+
+  // the size of rank limits how large associativity can be used in practice
+  __shared__ int rank[items_per_thread * nthreads];
+  rank_set_entries<nthreads, associativity>(cache_time, n_cache_sets, rank);
+
+  // Each thread will fill items_per_thread items in the cache.
+  // It uses a place, only if it was not udated at the current time step
```
[the remark this answers is not in the document]
Comment (tfeher) [reply cpp/include/raft/cache/cache_util.cuh:265]: Fixed.

File: cpp/include/raft/cache/cache_util.cuh [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+  // (cache_time != time).
+  // We rank the places acconding to the time stamp, least recently used
```
[the remark this answers is not in the document]
Comment (tfeher) [reply cpp/include/raft/cache/cache_util.cuh:267]: Fixed.

**Review by tfeher — COMMENTED** [review]
Thanks @cjnolet for the review! I have addressed the issues.

**Comment by cjnolet** [comment]
@tfeher, I just merged branch-0.19 into your branch in hopes maybe CI wasn't seeing some of the recent changes. I'm still at a loss for why these hangs are occurring in CI.

**Comment by cjnolet** [comment]
rerun tests

**Comment by cjnolet** [comment]
The failures appear to be related to whether a P100 is selected as the CI machine. I notice the tests never timeout when a V100 or A100 is selected.

**Comment by cjnolet** [comment]
rerun tests

**Comment by cjnolet** [comment]
@gpucibot merge

**Pull request merged**